documents/design/FirmwareRefactor.md

```diff
@@ -0,0 +1,189 @@
+
+Author: Dallas Downing
+
+## Motivation
+
+The firmware for IGVC has largely been neglected for the past few years, being an afterthought once more pressing problems have been addressed. However, we have reached a point where the basic code running on the mbed can no longer support the needs of the team. The goal of this firmware redesign is to tackle some of the main challenges with the firmware at last competition and make it easier to maintain and alter in the future. This will also (hopefully) motivate a new incentive to apply actual design decisions in making changes to the code later.
+
+Here is a short list of the problems that have been identified with the firmware and the mbed. Note that this is not a comprehensive list as many issues depend on the electrical system and the main computer. Most of the issues are explained later in the design section.
+
+All mbed code is contained in one `main.cpp` file, which makes understanding and editing the code difficult
+Asynchronous networking code (for communicating with the computer) runs in the same thread as “real-time” code (PID control)
+There are no errors messages sent from the mbed to the computer, making it difficult to see what the mbed is doing when it fails or encounters a problem
+Messages sent between the computer and the mbed follow the same format with many optional fields instead of using dedicated message formats for different commands
+The E-stop signal is treated as a normal digital input but should be an interrupt input to stop the robot as soon as possible
+
+There are also a number of new features being added to the electrical system that the mbed now must support. Part of the redesign was made keeping these features in mind. The specifics of their implementation has been left out of this document, but it should be clear how one might implement such features using the new firmware.
+A “virtual bumper” system using lidar lites that can warn and stop the robot preemptively before it hits an obstacle
+A diagnostics system with various sensors for detecting hardware issues with the robot 
+
+## Background
+
+The mbed is one of the most important components of the robot. Without it, the computer would just be a fancy computer and the rest of the robot would be a metal box filled with wires and silicon. The mbed is the bridge that connects the computer with the rest of the electrical system. This section will briefly go over what the mbed does to provide context for the upcoming design changes.
+
+Here are the main tasks that it performs:
+
+- Connects to the computer to receive PID values and motor commands (in the form of speeds for the left and right motor)
+- Performs PID calculations for bringing the motors to the right speeds
+- Receives signals from the encoders and keeps track of motor speeds and direction
+- Sends serial commands to the motor controller to properly drive the motors
+- Checks for when the e-stop is triggered and disables the motors when it is
```

Comment:
It also clears constants from the PID (notably I, and D accumulation). This is actually more important than setting the speed to be 0 since the e stop system prevents the motors from running anyway.

Reply:
Yes, I left that part out of the document for clarity for non-software people. If we write more concrete documentation about the mbed, then this should definitely be put there.

Reply:
Yes, I left that part out of the document for clarity for non-software people. If we write more concrete documentation about the mbed, then this should definitely be put there.